### PR TITLE
feat: expose detectors and separate detect and collect functions

### PR DESCRIPTION
**Pull request by Le0Developer**
See #143 for motivation.

Not sure if `getSources()` and `getDetectors()` should be kept, because they are no longer used.

### CODE REVIEW

**Comment by xnerhu** [comment]
LGTM, but it exposes `detectors` publicly, so we would be obligated to keep up with backward compatibility. @r-valitov @Finesse  wdyt

**Comment by Finesse** [comment]
> sending the components to the server and then running `detect()` there

@Le0Developer Did you really try to run `detect` on server? As I can see, many detectors use browser APIs, which aren't available in Node.js environment. What you want to achieve requires a deeper refactoring, to make sure the detectors are pure functions (i.e. use no data other that the input arguments, and use no browser API).

Also, could you please clarify why you export `detectors`? It would make sense if the exported `detect` function accepted detectors as an argument. Such way you could `detectors` as a base for your custom list of detectors.

> but it exposes `detectors` publicly, so we would be obligated to keep up with backward compatibility

@xnerhu Not necessarily. You may state in the docs that the detector list type is `Record<string, (components: ComponentDict) => DetectorResponse>, and the actual list of functions is outside of version control.

**Comment by Le0Developer** [comment]
> As I can see, many detectors use browser APIs, which aren't available in Node.js environment.

IIRC those mostly comprise of functions related to getting the browser kind etc, which shouldn't be hard to move.

> What you want to achieve requires a deeper refactoring

Yes, which I planned to do in a later PR, so it's easier to review.

> Such way you could `detectors` as a base for your custom list of detectors

This would require turning the `components` and `detectors` dicts into `AbtractComponentDict` and `AbstractDetectorDict` respectively, which might break typing? Will need to check.

**Comment by Le0Developer** [comment]
> but it exposes detectors publicly, so we would be obligated to keep up with backward compatibility

@xnerhu That whole export block is already documented as being out of semantic versioning:

https://github.com/fingerprintjs/BotD/blob/a1538504cf8dff08841fae3b7ab6c23b872adb2e/src/index.ts#L49-L50

File: src/sources/browser.ts
Comment (Le0Developer) [inline src/sources/browser.ts:4]: I don't like the naming, but I can't think of anything better currently.

File: src/sources/browser.ts
Comment (Finesse) [inline src/sources/browser.ts:12]: IMHO sources should be as atomic as possible to facilitate tree-shaking and simple detectors code. I.e. this source should be split into 4 sources.

@r-valitov @xnerhu What do you think?

File: src/sources/browser.test.ts
Comment (Le0Developer) [inline src/sources/browser.test.ts:1]: No clue how to properly test this.

File: src/detector.ts
Comment (Finesse) [inline src/detector.ts:3]: This creates [a circular dependency](https://github.com/fingerprintjs/BotD/actions/runs/6145435367/job/16673913120#step:8:9)

File: src/api.ts
Comment (Finesse) [inline src/api.ts:16]: I'd try to describe the dependency between the components' and detectors' types in TypeScript (the detectors expect a specific type of component dictionary)

**Review by Finesse — CHANGES_REQUESTED** [review]
> > What you want to achieve requires a deeper refactoring
> 
> Yes, which I planned to do in a later PR, so it's easier to review.

This may contradict an architectural decision made by @r-valitov and @xnerhu. The idea is to separate information sources from simple utility functions. I personally think the code state in the PR is good.

File: src/sources/browser.ts
Comment (Le0Developer) [reply src/sources/browser.ts:12]: [detectPluginsLengthInconsistency](https://github.com/fingerprintjs/BotD/blob/2056ef48dbfc4877cb4d2bbed40b2f9d2a1271a3/src/detectors/plugins_inconsistency.ts#L3) would need to check the state of 4 different components then 🤷

Comment (Finesse) [reply src/sources/browser.ts:12]: Worth it, I think. A small helper function can reduce the cumber.

BTW splitting allows sources to fail while affecting a smaller part of the algorithm.

File: src/sources/android.ts
Comment (Finesse) [inline src/sources/android.ts:5]: Idea: these new little files are not necessary. You can add `isAndroid` directly to the dictionary in `src/sources/index.ts`.

**Review by Finesse — COMMENTED** [review]
LGTM. I can say it's a high-quality PR, thank you.

The project maintainers will make the final decision about the architecture changes in this PR.

File: src/sources/android.ts
Comment (Le0Developer) [reply src/sources/android.ts]: I'm wondering if moving [getDocumentFocus](https://github.com/fingerprintjs/BotD/blob/main/src/utils/browser.ts#L119) out of utils into a source would make more sense.

Also `getMozAppearanceSupport`, `isIPad` and `isDesktopSafari` are unused apart from the debug info in the playground.

Comment (Finesse) [reply src/sources/android.ts]: I meant this:
```diff
- documentFocus: function hasDocumentFocus() {
-    return getDocumentFocus()
-  },
+ documentFocus: getDocumentFocus,
```

Comment (Le0Developer) [reply src/sources/android.ts]: 🤦 Yes of course.

Comment (Finesse) [reply src/sources/android.ts]: > I'm wondering if moving [getDocumentFocus](https://github.com/fingerprintjs/BotD/blob/main/src/utils/browser.ts?rgh-link-date=2023-09-11T12%3A49%3A24Z#L119) out of utils into a source would make more sense.
>
> Also getMozAppearanceSupport, isIPad and isDesktopSafari are unused apart from the debug info in the playground.

These are questions for @r-valitov and @xnerhu

**Review by r-valitov — APPROVED** [review]

**Review by xnerhu — APPROVED** [review]

**Comment by xnerhu** [comment]
@Le0Developer LGTM, thank you.

**Pull request merged**
